agent/provider/src/cli/whitelist.rs

```diff
@@ -0,0 +1,200 @@
+    let (added, skipped) = domain_patterns.add(add);
+    let domain_patterns: DomainPatterns = domain_patterns.into();
+    domain_patterns.save(&config.domain_whitelist_file)?;
+    if !added.is_empty() {
+        println_conditional(&config, "Added patterns:");
+        WhitelistTable::from(DomainPatterns { patterns: added }).print(&config)?
+    } else if config.json {
+        println_conditional(&config, "No new patterns to add.");
+        // no new pattern to add, so empty list for json output
+        WhitelistTable::from(DomainPatterns {
+            patterns: Vec::new(),
+        })
+        .print(&config)?
+    }
+    if !skipped.is_empty() && !config.json {
+        println_conditional(&config, "Dropped duplicated patterns:");
+        if !config.json {
+            WhitelistTable::from(DomainPatterns { patterns: skipped }).print(&config)?
+        }
+    }
+    Ok(())
+}
+
+fn remove(config: ProviderConfig, remove: Remove) -> anyhow::Result<()> {
+    let domain_patterns = DomainPatterns::load_or_create(&config.domain_whitelist_file)?;
+    let mut domain_patterns = DomainPatternIds::from(domain_patterns);
+    let (removed, _) = domain_patterns.remove(remove.ids);
+    let domain_patterns: DomainPatterns = domain_patterns.into();
+    domain_patterns.save(&config.domain_whitelist_file)?;
+    if !removed.is_empty() {
+        let table = WhitelistTable::from(DomainPatterns { patterns: removed });
+        println_conditional(&config, "Removed patterns:");
+        table.print(&config)?;
+    } else {
+        println_conditional(&config, "No matching pattern to remove.");
+        // no new pattern added, so empty list for json output
+        if config.json {
+            WhitelistTable::from(DomainPatterns {
+                patterns: Vec::new(),
+            })
+            .print(&config)?
+        }
+    };
+    Ok(())
+}
+
+struct WhitelistTable {
+    table: ResponseTable,
+}
+
+impl WhitelistTable {
+    pub fn new() -> Self {
+        let columns = vec!["ID".to_string(), "Pattern".to_string(), "Type".to_string()];
+        let values = vec![];
+        let table = ResponseTable { columns, values };
+        Self { table }
+    }
+
+    fn add(&mut self, pattern: DomainPattern) {
+        let id = pattern_to_id(&pattern);
+        let row = serde_json::json! {[ id, pattern.domain, pattern.domain_match ]};
+        self.table.values.push(row);
+    }
+
+    pub fn print(self, config: &ProviderConfig) -> anyhow::Result<()> {
+        let output = CommandOutput::from(self.table);
+        output.print(config.json)?;
+        Ok(())
+    }
+}
+
+impl From<DomainPatterns> for WhitelistTable {
+    fn from(domain_patterns: DomainPatterns) -> Self {
+        let mut table = WhitelistTable::new();
+        for pattern in domain_patterns.patterns {
+            table.add(pattern);
+        }
+        table
+    }
+}
+
+struct DomainPatternIds {
+    pattern_ids: HashMap<String, DomainPattern>,
+}
```

Comment:
Please add doc comments. It is not clear outright what the `String` type in `HashMap<String, DomainPattern>` does.

Reply:
Agree. I do not see any naming convention for maps in this project.
I would even make it a `type alias` if I would not need to implement `From` for it. It could also be a newtype idiom, but I hate seeing these `.0` everywhere.
Done.